### PR TITLE
Fix(?) universe handling in automatic obligation solving

### PR DESCRIPTION
**Pull request by SkySkimmer**
The added test broke with the version of #16995 which used the ugraph without this commit.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
Couldn't we sit down and write a spec at some point?

File: vernac/declare.ml [hunk 1]
```diff
@@ -2273,8 +2273,15 @@ let solve_and_declare_by_tac prg obls i tac =
...
     let poly = Internal.get_poly prg in
-    let prg = ProgramDecl.Internal.set_uctx ~uctx prg in
+    let uctx = if poly then uctx else UState.union prg.prg_uctx uctx in
     let def, obl', _cst = declare_obligation prg obl ~body:t ~types:ty ~uctx in
+    let uctx =
+      (* imitate obligation_terminator, XXX no idea if this is really correct *)
+      if poly then UState.union prg.prg_uctx uctx
+      else if def then UState.Internal.reboot (Global.env()) uctx
+      else uctx
+    in
```
Comment (ppedrot) [inline vernac/declare.ml:2283]: The `set_uctx` function is pure and we're calling it a couple of lines below with a similar dispatch on poly / def, can't we merge both calls?

Comment (ppedrot) [reply vernac/declare.ml:2283]: @SkySkimmer ping

Comment (SkySkimmer) [reply vernac/declare.ml:2283]: I don't know

Comment (ppedrot) [reply vernac/declare.ml:2283]: What do you mean, I don't know? You can clearly simplify the code.

Comment (SkySkimmer) [reply vernac/declare.ml:2284]: right, I forgot how this code looked

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]
I have no idea of the spec so this cannot be wronger.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**